### PR TITLE
fix(QTDI-2489): Apply default value before required validation for absent keys

### PR DESCRIPTION
**Pull request by thboileau**
### Requirements

* Any code change adding any logic **MUST** be tested through a unit test executed with the default build
* Any API addition **MUST** be done with a documentation update if relevant 

### Why this PR is needed?

Fixes [QTDI-2489](https://qlik-dev.atlassian.net/browse/QTDI-2489).

When a connector author adds a new `@Required + @DefaultValue` field to an existing connector, the TCK framework was throwing `IllegalArgumentException` because `@Required` validation fired on the raw config map **before** `@DefaultValue` was applied to absent keys. Existing serialized configurations (missing the new key) caused a false-positive validation failure.

**Root cause**: In `ReflectionService.PayloadValidator.onParameter`, the required-field check compared the raw value against `JsonValue.NULL` without first checking whether a default was declared. For newly added fields absent from a stored config, `PayloadMapper` delivers `JsonValue.NULL` — the validation fired even though the declared default would have been applied moments later.

### What does this PR add (design/code thoughts)?

**One-line guard** in `PayloadValidator.onParameter`:

```java
// BEFORE
if (Boolean.parseBoolean(meta.getMetadata().get("tcomp::validation::required"))
        && value == JsonValue.NULL) {
    errors.add(MESSAGES.required(meta.getPath()));
}

// AFTER
if (Boolean.parseBoolean(meta.getMetadata().get("tcomp::validation::required"))
        && value == JsonValue.NULL
        && meta.getMetadata().get("tcomp::ui::defaultvalue::value") == null) {
    errors.add(MESSAGES.required(meta.getPath()));
}
```

The `tcomp::ui::defaultvalue::value` metadata key is set by `UiParameterEnricher` when `@DefaultValue` is present. When it is non-null, we suppress the required-field error — the default will be applied later in the object factory chain. When it is absent, the existing behaviour is preserved: a genuinely missing required field still throws.

**Tests added** (5 new tests in `ReflectionServiceTest`):

| Test | What it verifies |
|---|---|
| `validationRequiredWithDefaultValueMissingKeyPasses` | Missing key + `@DefaultValue` → no exception |
| `validationRequiredWithDefaultValueResolvedValue` | Missing key + `@DefaultValue("DEFAULT")` → resolved value equals `"DEFAULT"` |
| `validationRequiredNoDefaultValueMissingKeyFails` | Missing key + no `@DefaultValue` → `IllegalArgumentException` (regression) |
| `validationRequiredWithEmptyDefaultValuePasses` | Missing key + `@DefaultValue("")` → no exception |
| `validationRequiredMultipleAbsentDefaultsPasses` | Multiple missing keys all with defaults → no exception |

All 50 tests in `ReflectionServiceTest` pass (45 pre-existing + 5 new).

### AI generated code
https://internal.qlik.dev/general/ways-of-working/code-reviews/#guidelines-for-ai-generated-code
- [x] this PR has been written with the help of GitHub Copilot or another generative AI tool

[QTDI-2489]: https://qlik-dev.atlassian.net/browse/QTDI-2489?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by thboileau** [comment]
## 🔍 Scope & Design Review — QTDI-2489 (Round 1)

**Verdict: APPROVED** — zero Blocker/Major findings.

| Finding | Severity | Detail |
|---|---|---|
| Metadata key `tcomp::ui::defaultvalue::value` | None | Verified from `UiParameterEnricher` and `StreamingLongParamBuilder` |
| Explicit-null vs absent-key | None | `PayloadMapper` delivers `JsonValue.NULL` only for absent map keys; explicitly null entries don't occur in `Map<String,String>` configs |
| AC1 — default suppresses required error | None | Guard correctly fires when required AND absent AND defaultvalue declared |
| AC2 — present key path unchanged | None | `value != JsonValue.NULL` bypasses guard entirely |
| AC3 — genuinely missing required field still fails | None | `@DefaultValue` absent → `get(...) == null` is true → error fires as before |
| AC4 — no migration handler required | None | Fix confined to validation layer; no connector change needed |
| `@DefaultValue` on complex types | Info | Annotation is primitive-only per Javadoc; TCK maven plugin enforces at build time — no runtime risk |
| Test coverage | None | 5 new tests cover all ACs plus empty-string and multi-absent edge cases |

**Fixes applied in Round 1**: None.

---
*Review performed by AI (GitHub Copilot). Human reviewer should verify the guard condition and test coverage.*

**Comment by thboileau** [comment]
## ✅ Compliance Check — QTDI-2489

**Verdict: PASS** — no compliance violations.

### Files reviewed

1. `component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java`
2. `component-runtime-manager/src/test/java/org/talend/sdk/component/runtime/manager/ReflectionServiceTest.java`

### Findings

| Rule | File | Severity | Status |
|---|---|---|---|
| Spotless formatting | Both | — | PASS — auto-applied before commit |
| Checkstyle | Both | — | PASS — 0 violations |
| No wildcard imports | Test | — | PASS — specific import added |
| JUnit 5 — no `public` on `@Test` | Test | — | PASS — default visibility on all 5 new methods |
| Scope discipline | Both | — | PASS — all changes traceable to approved plan |
| Test level (UT) | Test | — | PASS — pure framework logic, UT correct |
| Test scope coverage | Test | — | PASS — all 5 items from test scope covered |
| No deprecated API usage | Both | — | PASS — no deprecated APIs called |
| No opportunistic refactoring | Both | — | PASS — minimal required change only |

### Severity summary

| Severity | Count |
|---|---|
| Critical fixed | 0 |
| Warnings | 0 |
| Info | 0 |

File: component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java [hunk 1]
```diff
@@ -913,95 +913,125 @@ public void onParameter(final ParameterMeta meta, final JsonValue value) {
...
             final Map<String, String> metadata = meta.getMetadata();
-            {
-                final String min = metadata.get("tcomp::validation::min");
-                if (min != null) {
-                    final double bound = Double.parseDouble(min);
-                    if (value.getValueType() == JsonValue.ValueType.NUMBER
-                            && ((JsonNumber) value).doubleValue() < bound) {
-                        errors.add(MESSAGES.min(meta.getPath(), bound, ((JsonNumber) value).doubleValue()));
-                    }
+            validateRuleMin(meta, value, metadata);
```
Comment (thboileau) [inline component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java:920]: Following Sonar comment: reduce method complexity. Each block has been extracted as a method

**Comment by thboileau** [comment]
Issue with Sonar -> https://qlik-dev.atlassian.net/browse/QTDI-2489?focusedCommentId=1172248
let's wait for the next release of TCK (should reset the base reference of Sonar), and rerun Sonar CI then.

**Comment by thboileau** [comment]
The release of tCK did not change the Sonar report: it still complains that ReflectionService.PayloadValidator.validate* methods are not tested.

image



But, when we have a look at the [jacoco report](https://jenkins-connectors.rnd-ci.cloud.talend.com/job/component-runtime/job/build/job/PR-1254/11/artifact/reporting/target/site/jacoco-aggregate/component-runtime-manager/org.talend.sdk.component.runtime.manager.reflect/ReflectionService$PayloadValidator.html), we see that these methods are tested.

image

File: component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java [hunk 2]
```diff
@@ -913,95 +913,123 @@ public void onParameter(final ParameterMeta meta, final JsonValue value) {
                 return;
             }
 
-            if (Boolean.parseBoolean(meta.getMetadata().get("tcomp::validation::required"))
-                    && value == JsonValue.NULL) {
+            if (isRequiredAndMissingWithoutDefault(meta, value)) {
```
Comment (thboileau) [inline component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java:916]: Here is the fix: required rule fails if there is no default value

File: component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java [hunk 3]
```diff
@@ -913,95 +913,123 @@ public void onParameter(final ParameterMeta meta, final JsonValue value) {
...
+            if (min != null) {
+                final double bound = Double.parseDouble(min);
+                if (value.getValueType() == JsonValue.ValueType.NUMBER
+                        && ((JsonNumber) value).doubleValue() < bound) {
+                    errors.add(MESSAGES.min(meta.getPath(), bound, ((JsonNumber) value).doubleValue()));
                 }
             }
         }
 
+        private static boolean isRequiredAndMissingWithoutDefault(final ParameterMeta meta,
```
Comment (thboileau) [inline component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java:1026]: It lacks one use case: when the default value is an empty string. In such case, the default value is considered as irrelevant

```
    @Option
    @Required
    @DefaultValue("")
    private String fieldWithEmptyDefault;
```

File: component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java (hunk 1 above)
Comment (thboileau) [reply component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java:920]: Acknowledged — this is a self-annotation explaining the Round 2/4 Sonar-driven refactor (cognitive complexity extraction). No further code change needed here.

File: component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java (hunk 2 above)
Comment (thboileau) [reply component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java:916]: Acknowledged — this is a self-annotation explaining the original Round 0 guard. No further code change needed here.

File: component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java (hunk 3 above)
Comment (thboileau) [reply component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/reflect/ReflectionService.java:1026]: Fixed in `55b3803f435` — `isRequiredAndMissingWithoutDefault` now also requires the declared default value to be non-empty, so `@DefaultValue("")` no longer suppresses the required-field error for an absent key. Added `validationRequiredWithEmptyDefaultValueMissingKeyFails` to cover it.

**Comment by thboileau** [comment]
## Review round N — summary

| Comment | Class | Action |
|---|---|---|
| #1 (2026-07-15, line 920) — "Following Sonar comment: reduce method complexity..." | Already addressed | Replied — self-annotation, no code change needed |
| #2 (2026-07-28, line 916) — "Here is the fix: required rule fails if there is no default value" | Already addressed | Replied — self-annotation, no code change needed |
| #3 (2026-08-31, line 1026) — "It lacks one use case: when the default value is an empty string..." | Code fix | Fixed in `55b3803f435` |

**Fixes pushed**: `55b3803f435` — fix(QTDI-2489): address review round N — empty-string default no longer suppresses required validation

**Build**: 64/64 tests pass in `ReflectionServiceTest` — GREEN.

**Rebase**: no — `master` had not moved since last push.

_Round summary generated by AI. Please resolve threads after verifying the fixes._
